### PR TITLE
Migrate pp system implementation to osuAkatsuki/akatsuki-pp-rs

### LINKED ISSUE #101: Add other gamemodes (taiko, ctb, mania) to !recalc commands
that's useful to recalc pp in all maps to live test changes made in oppai-ng <3

### LINKED ISSUE #184: Write separate pp recalculation script
we can use the new API provided in #177

it could also be ran via subprocess from a command, to implement something alike !recalc

### LINKED ISSUE #283: Migrate pp system implementation to rosu-pp-py
rosu-pp is a high performance rust implementation of the pp system with support for all osu! modes, which recently also added support for converts.

implementing this will both simplify our pp calculation logic (we currently rely on two separate libraries in different language), as well as enhancing support for the minority game modes

https://pypi.org/project/rosu-pp-py
https://github.com/MaxOhn/rosu-pp-py

### PR DESCRIPTION
**Pull request by 7mochi**
Closes #283, closes #184, closes #101

### CODE REVIEW

**Comment by minisbett** [comment]
Why would you make the AR factor 0 on relax

**Comment by minisbett** [comment]
This is the worst relax implementation I've ever seen

**Comment by minisbett** [comment]
Nvm it's the bancho implementation I forgot it sucks

**Comment by minisbett** [comment]
We will definitely have to fork rosupp i don't think anyone here wants 1k pp plays getting nerfed to 200-300pp

**Comment by tsunyoku** [comment]
tbh just use https://github.com/osuAkatsuki/akatsuki-pp-rs since it's rosu-pp too.
just note the relax system lies in osu_2019 and not osu, because adapting 2022 for relax was awful

i'd also like to note that rosu_pp_py is not up to date with the latest rosu-pp version (due to rewrite stuff), so i don't think making this change is a good idea right now

**Comment by minisbett** [comment]
Also better for customization so for this PR I would suggest putting in rosupp-py as a submodule instead

**Comment by minisbett** [comment]
> tbh just use https://github.com/osuAkatsuki/akatsuki-pp-rs since it's rosu-pp too. just note the relax system lies in osu_2019 and not osu, because adapting 2022 for relax was awful
> 
> i'd also like to note that rosu_pp_py is not up to date with the latest rosu-pp version (due to rewrite stuff), so i don't think making this change is a good idea right now

Or that would work

**Comment by 7ez** [comment]
just replace vn pp with rosu-pp, I don't see a sense in replacing the whole std calc with it

**Comment by tsunyoku** [comment]
> just replace vn pp with rosu-pp, I don't see a sense in replacing the whole std calc with it

because we get to use a single system across the board which makes life much easier
also oppai is problematic & peace is outdated and doesn't support converts

however, we need to wait for rosu-pp-py to be updated at the very least; unless we decide to just use akatsuki's version before then

**Comment by 7mochi** [comment]
> tbh just use https://github.com/osuAkatsuki/akatsuki-pp-rs since it's rosu-pp too. just note the relax system lies in osu_2019 and not osu, because adapting 2022 for relax was awful
> 
> i'd also like to note that rosu_pp_py is not up to date with the latest rosu-pp version (due to rewrite stuff), so i don't think making this change is a good idea right now

I could try adapting that version with the latest rosu-pp changes. Would that work?

**Comment by tsunyoku** [comment]
> > tbh just use https://github.com/osuAkatsuki/akatsuki-pp-rs since it's rosu-pp too. just note the relax system lies in osu_2019 and not osu, because adapting 2022 for relax was awful
> > i'd also like to note that rosu_pp_py is not up to date with the latest rosu-pp version (due to rewrite stuff), so i don't think making this change is a good idea right now
> 
> I could try adapting that version with the latest rosu-pp changes. Would that work?

it already includes them, it’s up to date with rosu-pp’s main branch

**Comment by minisbett** [comment]
Either way, I still suggest putting rosupp-py in as a submodule so it can be modified

**Comment by cmyui** [comment]
nice work so far, thanks for taking this responsibility on!

**Comment by cmyui** [comment]
also, looks like i must've broke pre-commit ci at some point 🤔

**Comment by 7mochi** [comment]
> > > tbh just use https://github.com/osuAkatsuki/akatsuki-pp-rs since it's rosu-pp too. just note the relax system lies in osu_2019 and not osu, because adapting 2022 for relax was awful
> > > i'd also like to note that rosu_pp_py is not up to date with the latest rosu-pp version (due to rewrite stuff), so i don't think making this change is a good idea right now
> > 
> > 
> > I could try adapting that version with the latest rosu-pp changes. Would that work?
> 
> it already includes them, it’s up to date with rosu-pp’s main branch

Which python bindings would I use with akatsuki-pp-rs?

**Comment by cmyui** [comment]
> Which python bindings would I use with akatsuki-pp-rs?

@tsunyoku if you could help him out a bit here would be very epic 11/10

**Comment by 7mochi** [comment]
> > Which python bindings would I use with akatsuki-pp-rs?
> 
> @tsunyoku if you could help him out a bit here would be very epic 11/10

akatsuki-pp-rs doesn't have python bindings yet so idk know how I should proceed 💀
I guess I'll wait until there is one because I haven't been able to adapt those changes to rosu-pp-py on my own

**Comment by minisbett** [comment]
> > > Which python bindings would I use with akatsuki-pp-rs?
> > 
> > 
> > @tsunyoku if you could help him out a bit here would be very epic 11/10
> 
> akatsuki-pp-rs doesn't have python bindings yet so idk know how I should proceed 💀 I guess I'll wait until there is one because I haven't been able to adapt those changes to rosu-pp-py on my own

Since akatsuki-pp-rs is based on rosu-pp shouldn't the normal rosu-pp bindings work? Idk how python bindings work but if I imagine it correct it's basically just saying access these and these methods and the methods should have the same signatures

**Comment by 7mochi** [comment]
I think I got it to work, but there's no bindings repo in osuAkatsuki so i can't make a PR yet, anyways, i've tested it in my fork and it works. Could you guys check it? @cmyui @tsunyoku 

Changes:
- rosu-pp-py: https://github.com/osuperu/rosu-pp-py/commit/70329714e7f5e3bcb402ba985540a154da15f0ad
- akatsuki-pp-rs: https://github.com/osuperu/rosu-pp/commit/cd002b0c3e7228058684b8e3345258ff8efb5a9f
- bancho.py (performance.py | calculate_performances_std)
```py
if score["mods"] and Mods.RELAX & score["mods"]:
    calculator.set_acc_2019(acc)
    result = calculator.performance_2019(map)
else:
    result = calculator.performance(map)
```

**Comment by cmyui** [comment]
nice, @tsunyoku is working on making bindings for akatsuki's pp system

**Comment by cmyui** [comment]
i'll review this shortly

**Comment by cmyui** [comment]
@7mochi hey, do you have any uncommitted changes? i'd like to take over the pr for a bit when you're ready

**Comment by 7mochi** [comment]
> @7mochi hey, do you have any uncommitted changes? i'd like to take over the pr for a bit when you're ready

Just that piece of code I posted earlier

**Comment by tsunyoku** [comment]
hey, made bindings for akatsuki-pp-rs @ https://github.com/osuAkatsuki/akatsuki-pp-py.
all tested including automatic osu_2019 usage when osu!std relax is passed (use the same api as rosu-pp-py examples)

hope this helps :D

File: README_CN.md [hunk 1]
```diff
@@ -81,6 +82,9 @@ sudo apt install -y python3.9-dev python3.9-distutils \
                     mysql-server redis-server \
                     nginx certbot
 
+# 安装rust
+curl https://sh.rustup.rs -sSf | bash -s -- -y
```
Comment (cmyui) [inline README_CN.md:86]:
```suggestion
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```

File: README_DE.MD [hunk 2]
```diff
@@ -61,6 +62,9 @@ sudo apt install -y python3.9-dev python3.9-distutils \
                     mysql-server redis-server \
                     nginx certbot
 
+# Installieren rust
+curl https://sh.rustup.rs -sSf | bash -s -- -y
```
Comment (cmyui) [inline README_DE.MD:66]:
```suggestion
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```

File: README.md [hunk 3]
```diff
@@ -58,6 +59,9 @@ sudo apt install -y python3.9-dev python3.9-distutils \
                     mysql-server redis-server \
                     nginx certbot
 
+# install rust
+curl https://sh.rustup.rs -sSf | bash -s -- -y
```
Comment (cmyui) [inline README.md:63]:
```suggestion
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```

File: app/commands.py [hunk 4]
```diff
@@ -1288,24 +1299,68 @@ async def recalc_all() -> None:
...
+                                if mode in (
+                                    GameMode.VANILLA_TAIKO,
+                                    GameMode.RELAX_TAIKO,
+                                ):
+                                    modeInt = GameMode.VANILLA_TAIKO
+                                elif mode in (
+                                    GameMode.VANILLA_CATCH,
+                                    GameMode.RELAX_CATCH,
+                                ):
+                                    modeInt = GameMode.VANILLA_CATCH
```
Comment (tsunyoku) [inline app/commands.py:1333]:
```suggestion
                                    mode_int = GameMode.VANILLA_CATCH
```

File: app/commands.py [hunk 5]
```diff
@@ -1288,24 +1299,68 @@ async def recalc_all() -> None:
...
+                                GameMode.VANILLA_CATCH,
+                                GameMode.RELAX_CATCH,
                             ):
-                                ezpp.set_mods(row["mods"])
-                                ezpp.set_nmiss(row["nmiss"])  # clobbers acc
-                                ezpp.set_combo(row["max_combo"])
-                                ezpp.set_accuracy_percent(row["acc"])
 
-                                ezpp.calculate(str(osu_file_path))
+                                modeInt = GameMode.VANILLA_OSU
```
Comment (tsunyoku) [inline app/commands.py:1323]:
```suggestion
                                mode_int = GameMode.VANILLA_OSU
```

File: app/commands.py [hunk 6]
```diff
@@ -1288,24 +1299,68 @@ async def recalc_all() -> None:
...
+                                    GameMode.RELAX_TAIKO,
+                                ):
+                                    modeInt = GameMode.VANILLA_TAIKO
+                                elif mode in (
+                                    GameMode.VANILLA_CATCH,
+                                    GameMode.RELAX_CATCH,
+                                ):
+                                    modeInt = GameMode.VANILLA_CATCH
+
+                                calculator = Calculator(mods=row["mods"], mode=modeInt)
```
Comment (tsunyoku) [inline app/commands.py:1335]:
```suggestion
                                calculator = Calculator(mods=row["mods"], mode=mode_int)
```

Comment (tsunyoku) [inline app/commands.py:1335]:
```suggestion
                                calculator = Calculator(mods=row["mods"], mode=mode_int)
```

**Review by tsunyoku — CHANGES_REQUESTED** [review]

**Comment by cmyui** [comment]
going to refactor the calculate_performances_x functions into a singular interface, now that we can

File: app/constants/gamemodes.py [hunk 7]
```diff
@@ -59,6 +59,20 @@ def from_params(cls, mode_vn: int, mods: Mods) -> GameMode:
...
+        ret = []
+        for mode in cls:
+            if mode not in (
+                cls.RELAX_MANIA,
+                cls.AUTOPILOT_TAIKO,
+                cls.AUTOPILOT_CATCH,
+                cls.AUTOPILOT_MANIA,
+            ):
+                ret.append(mode)
+        return ret
```
[the remark this answers is not in the document]
Comment (cmyui) [reply app/constants/gamemodes.py:74]: thought about this one, will keep as is.

property seems more appropriate for things on the instances, e.g. GameMode(0).as_vanilla, whereas this is meant to be used on the class itself, GameMode.valid_gamemodes()

memory leak only applies to methods, not classmethods

**Comment by cmyui** [comment]
getting close!

**Comment by cmyui** [comment]
I think we're good to go on this.

Can I get a re-review @tsunyoku, @7mochi please!

**Comment by cmyui** [comment]
also @alowave223 you should take a look, i think

**Comment by cmyui** [comment]
ah, found some bugs with callers of `calculate_performances` still using params that were moved into the `scores` param

**Comment by cmyui** [comment]
ok, things look good now. feel free to review away!

**Review by cmyui — APPROVED** [review]

**Review by tsunyoku — APPROVED** [review]

**Pull request merged**